applications/asset_tracker_v2/src/modules/location_module.c

```diff
@@ -0,0 +1,682 @@
+	 * to let the GNSS module finish ongoing searches before data is sent to cloud.
+	 */
+	if (methods_index_gnss != -1) {
+		config.methods[methods_index_gnss].gnss.timeout =
+			MAX(copy_cfg.gnss_timeout + 15, 75) * MSEC_PER_SEC;
+	}
+	if (methods_index_cellular != -1) {
+		config.methods[methods_index_cellular].cellular.timeout =
+			DATA_FETCH_TIMEOUT_NEIGHBORHOOD_SEARCH * MSEC_PER_SEC;
+	}
+
+	LOG_DBG("Requesting location...");
+
+	err = location_request(&config);
+	if (err) {
+		// TODO:
+		// SEND_EVENT(modem, LOCATION_EVT_LOCATION_DATA_NOT_READY);
+		LOG_ERR("Location request failed: %d", err);
+		return;
+	}
+	SEND_EVENT(location, LOCATION_MODULE_EVT_ACTIVE);
+	stats.start_uptime = k_uptime_get();
+}
+
+static void inactive_send(void)
+{
+	SEND_EVENT(location, LOCATION_MODULE_EVT_INACTIVE);
+}
+
+static void time_set(void)
+{
+	/* Change datetime.year and datetime.month to accommodate the
+	 * correct input format.
+	 */
+	struct tm gnss_time = {
+		.tm_year = pvt_data.datetime.year - 1900,
+		.tm_mon = pvt_data.datetime.month - 1,
+		.tm_mday = pvt_data.datetime.day,
+		.tm_hour = pvt_data.datetime.hour,
+		.tm_min = pvt_data.datetime.minute,
+		.tm_sec = pvt_data.datetime.seconds,
+	};
+
+	date_time_set(&gnss_time);
+}
+
+static inline int adjust_rsrp(int input)
+{
+	if (IS_ENABLED(CONFIG_MODEM_NEIGHBOR_CELLS_DATA_CONVERT_RSRP_TO_DBM)) {
+		return RSRP_IDX_TO_DBM(input);
+	}
+
+	return input;
+}
+
+static inline int adjust_rsrq(int input)
+{
+	if (IS_ENABLED(CONFIG_MODEM_NEIGHBOR_CELLS_DATA_CONVERT_RSRQ_TO_DB)) {
+		return round(RSRQ_IDX_TO_DB(input));
+	}
+
+	return input;
+}
+
+static void send_neighbor_cell_update(struct lte_lc_cells_info *cell_info)
+{
+	struct modem_module_event *evt = new_modem_module_event();
+
+	BUILD_ASSERT(sizeof(evt->data.neighbor_cells.cell_data) ==
+		     sizeof(struct lte_lc_cells_info));
+	BUILD_ASSERT(sizeof(evt->data.neighbor_cells.neighbor_cells) >=
+		     sizeof(struct lte_lc_ncell) * CONFIG_LTE_NEIGHBOR_CELLS_MAX);
+
+	memcpy(&evt->data.neighbor_cells.cell_data, cell_info, sizeof(struct lte_lc_cells_info));
+	memcpy(&evt->data.neighbor_cells.neighbor_cells, cell_info->neighbor_cells,
+	       sizeof(struct lte_lc_ncell) * evt->data.neighbor_cells.cell_data.ncells_count);
+
+	/* Convert RSRP to dBm and RSRQ to dB per "nRF91 AT Commands" v1.7. */
+	evt->data.neighbor_cells.cell_data.current_cell.rsrp =
+			adjust_rsrp(evt->data.neighbor_cells.cell_data.current_cell.rsrp);
+	evt->data.neighbor_cells.cell_data.current_cell.rsrq =
+			adjust_rsrq(evt->data.neighbor_cells.cell_data.current_cell.rsrq);
+
+	for (size_t i = 0; i < evt->data.neighbor_cells.cell_data.ncells_count; i++) {
+		evt->data.neighbor_cells.neighbor_cells[i].rsrp =
+			adjust_rsrp(evt->data.neighbor_cells.neighbor_cells[i].rsrp);
+		evt->data.neighbor_cells.neighbor_cells[i].rsrq =
+			adjust_rsrq(evt->data.neighbor_cells.neighbor_cells[i].rsrq);
+	}
+
+	evt->type = MODEM_EVT_NEIGHBOR_CELLS_DATA_READY;
```

Comment:
This function of sending neighbor cell data (also adjust_rsrq and adjust_rsrq) has been copied from modem module as is.
We can think of renaming the event `MODEM_EVT_NEIGHBOR_CELLS_DATA_READY` to `LOCATION_EVT_NEIGHBOR_CELLS_DATA_READY`. and converting this event into location module event.

I've first though neightbor cell stuff would be modem module's responsibility but maybe I'll just convert this handling into location events. Opinions?

Reply:
I would prefer having this in the location module. So that everything that is related to location comes from the same source. I think that very justifiable. In my initial comment on this PR I suggested having a single event
for location and providing the data in an union. So that GNSS fix, ncell meas and wifi (at some point) is carried in the same LOCATION_EVT_LOCATION_DATA_READY event.